packages/transformers/js/core/src/utils.rs

```diff
@@ -168,6 +168,34 @@ pub fn match_require(node: &ast::Expr, unresolved_mark: Mark, ignore_mark: Mark)
   }
 }
 
+/// This matches an expression like `importDeferredForDisplay('id')` or `importDeferred('id')` and returns the dependency id.
```

Comment:
Can you use [import attributes](https://github.com/tc39/proposal-import-attributes) for this?

```js
import('id', {with: {tier: 'display'}})
```

This syntax is now supported by TS and other tooling as well so should be safe to use. The advantage over individual functions is that the priorities aren't hard coded into the transformer, and no special TS declarations for these new non-standard global functions would be needed. Internally, you'd also not need any special dependency types or changes to core data structures, and could just use the import attributes stored on the dependency for the metadata.

Reply:
I evaluated this earlier because it would definitely be easier to ship. Unfortunately we're unable to use it yet as there's no way to safely type it (https://github.com/microsoft/TypeScript/issues/46135). I also saw some issues pop up in Jira with our tooling crashing out which made it more time consuming to implement (SWC wasn't happy, nor was Jest).

I'm keen to migrate over though when the type support it added

Reply:
Maybe I misunderstood the purpose of this feature, but why does the type of the returned module need to be different from a normal import? I assumed it was just about prioritizing the load order.

Reply:
I'll give a quick overview for the motivation and approach for the project:

Tiers are a concept that allows us to reason about code in different stages of the page load so that developers can have full control over when the user is loading and executing their code. Web vitals give us guidance on what stages we need to consider for user experience, and we can have corresponding tiers.

First is the “loading” tier, where we need to make clear to the users that we’re working on loading their content. We already achieved this pattern with “skeletons”, but we can make this much more efficient by making a specific tier for this and showing the loading state as early as possible. This corresponds to sync imports.

Second is the “visually complete” tier, where the focus is on ensuring users can see everything they need and start interacting with the experience. This tier can be considered SSR returning most of the page as well as any React code required to enable basic interactivity in the page, such that users can make interactions that are actioned when the required code has loaded. This corresponds to `importDeferredForDisplay`.

Third is the “fully interactive” state, where users can interact fully with the experience without waiting for deferred content to appear. Less critical code, such as logging and background tasks, should be executed in this tier. Importantly, as the visually complete tier has already happened, we should not trigger a paint. We should expect no visual changes during this tier and a Cumulative Layout Shift of ~0. This corresponds to `importDeferred`.

These new imports need to return a different type so we can treat them differently at runtime, especially for using stuff like suspense (like in the example). Stuff like react-loosely-lazy is what we used before, but this isn't as efficient nor bundler aware. Each tier we want to have a stage of "blocked" and "unblocked" so that each tier cannot interfere with another, which happens with react-loosely-lazy. That's important because in our current solution, we end up with microtasks from critical code being run after non-critical code (such as stuff that doesn't paint or just stuff like analytics).

Reply:
I'm still not sure why that would require a special type. Couldn't the priority queue be internal to the runtime and not exposed to user code?

Reply:
The main issue we have here is we want sync loading if the module is available and promise loading if it's not (like https://relay.dev/docs/glossary/#observable). The APIs available to us right now don't allow this, it's one or the other (`import x from 'y'` or `x = import('y')`). Happy to chat about this in more depth in the core catch up

Reply:
I believe React supports this out of the box using some extensions to the Promise object, namely `status`, `value`, and `reason`. See the RFC for `use()` here: https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises.md#reading-the-result-of-a-promise-that-was-read-previously.

Maybe you can use something similar? It would be really nice to avoid adding new APIs.